pkgng: Add support for upgrading all installed packages

dsg22 opened the pull request:
##### SUMMARY
Adds support for ``name: "*", state: latest`` to upgrade all installed packages, similar to other package providers.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
pkgng

##### ADDITIONAL INFORMATION

Fixes ansible/ansible#35515

felixfontein commented on changelogs/fragments/569-pkgng-add-upgrade-action.yaml:
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - Add an "upgrade" parameter to the pkgng module to upgrade all installed packages
```
```suggestion
  - pkgng - add an ``upgrade`` parameter to upgrade all installed packages (https://github.com/ansible-collections/community.general/pull/569).
```

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -23,14 +23,20 @@
+            - Upgrade all installed packages.
+        type: bool
+        required: false
+        default: no
```
```suggestion
        default: no
        version_added: 1.0.0
```

dsg22 replied:
Hmm, the module isn't really versioned right now AFAICT... do I need to add that as well?

felixfontein replied:
The collection is versioned, not the module. And the next collection release will be 1.0.0, so this option will show up in community.general 1.0.0.

dsg22 replied:
Understood, thanks for the clarification. Rebased and applied just now.

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -335,7 +363,8 @@ def main():
         argument_spec=dict(
             state=dict(default="present", choices=["present", "latest", "absent"], required=False),
-            name=dict(aliases=["pkg"], required=True, type='list'),
+            name=dict(aliases=["pkg"], default=[], required=False, type='list'),
```
Do I understand it correctly that `state` and `name` are ignored if `upgrade` is `True`? This needs to be properly documented.

Also, I would make `name` required if `upgrade` is not `True`.

dsg22 replied:
> Do I understand it correctly that `state` and `name` are ignored if `upgrade` is `True`? This needs to be properly documented.

No, `state` and `name` can be passed along with `upgrade`. The module will first perform the upgrade and then make state changes to packages.

> 
> Also, I would make `name` required if `upgrade` is not `True`.

Sure, I'll prepare a patch for that in the morning, leaving work now.

dsg22 replied:
I'm actually not sure how to achieve this with the provided checks in AnsibleModule. I'd need to make ``name`` required if both of ``upgrade`` or ``autoremove`` are false.

But after thinking it through, I don't think it's needed. The module can run with any combination of ``name``, ``upgrade`` and ``autoremove``, including none or all of them. If none are passed, the module will do nothing. What do you think?

Also, just wanted to say thanks for your help with this and the other PR that got merged yesterday. The PR process has been an absolute pleasure, it's great to see a project so welcoming of new contributions!

felixfontein replied:
If something is not supported, you can always do it manually (check and use `module.fail_json()`).

But if all combinations are ok, it's fine for me too. (Doing nothing by default is acceptable IMO.)

About the review process: let me mention that I'm not familiar with this module (or pkgng), so I can't provide a complete review on this one :) I hope that one of the maintainers can take a look at this one too...

dsg22 commented:
Squashed and rebased on current master

dsg22 commented:
After discussion in #ansible-devel the consensus seems to be that ``name: "*", state: latest`` is the preferred interface for this. I've reworked the patch to use that and updated the description to match.

felixfontein commented on changelogs/fragments/569-pkgng-add-upgrade-action.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+minor_changes:
+  - pkgng: Added support for upgrading all packages using ``name: *, state: latest``, similar to other package providers (https://github.com/ansible-collections/community.general/pull/569).
```
```suggestion
  - pkgng - added support for upgrading all packages using ``name: *, state: latest``, similar to other package providers (https://github.com/ansible-collections/community.general/pull/569).
```
Same for the others.

dsg22 replied:
Updated, thanks!

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -22,7 +22,7 @@
     name:
         description:
-            - Name or list of names of packages to install/remove.
+            - Name or list of names of packages to install/remove. With I(name=*), I(state: latest) will operate, but I(state: present) and I(state: absent) will be noops.
```
```suggestion
            - Name or list of names of packages to install/remove.
            - With I(name=*), I(state: latest) will operate, but I(state: present) and I(state: absent) will be noops.
```
I think with a new paragraph this will be easier to read.

dsg22 commented:
Fixed lint errors due to invalid YAML flagged by @ansibullbot above

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -391,18 +415,28 @@ def main():
 
-    if p["state"] in ("present", "latest"):
-        _changed, _msg, _stdout, _stderr = install_packages(module, pkgng_path, pkgs, p["cached"], p["pkgsite"], dir_arg, p["state"])
+    if '*' in pkgs and p["state"] == 'latest':
```
```suggestion
    if pkgs == ['*'] and p["state"] == 'latest':
```

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -391,18 +415,28 @@ def main():
 
+    # Operate on named packages
+    named_packages = [pkg for pkg in pkgs if pkg != '*']
+    if p["state"] in ("present", "latest"):
```
```suggestion
    if p["state"] in ("present", "latest") and named_packages:
```

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -391,18 +415,28 @@ def main():
+        changed = changed or _changed
+        msgs.append(_msg)
+
     elif p["state"] == "absent":
```
```suggestion
    elif p["state"] == "absent" and named_packages:
```

dsg22 commented:
See the last commit, I added a warning about breakage on older module versions with the upgrade syntax as it's quite bad. I think we need to call this out more prominently than in the changelog. I'd appreciate a review of the wording.

dsg22 commented:
Squashed all commits, no changes in this last push.

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -111,6 +118,11 @@
     name: baz
     state: latest
+
+- name: Upgrade all installed packages (see warning for I(name) first!)
```
That formatting doesn't work in `EXAMPLES`.
```suggestion
- name: Upgrade all installed packages (see warning for name first!)
```
(Maybe it's better to write `see warning for the name option first!` to make clear that the `name` option is meant.)

felixfontein commented on plugins/modules/packaging/os/pkgng.py:
```diff
@@ -23,6 +23,13 @@
             - Name or list of names of packages to install/remove.
+            - "With I(name=*), I(state: latest) will operate, but I(state: present) and I(state: absent) will be noops."
+            - >
+                Warning: In ansible 2.9 and earlier this module had a misfeature
```
```suggestion
                Warning: In Ansible 2.9 and earlier this module had a misfeature
```
Sorry for not noticing that earlier!

felixfontein commented:
@dsg22 thank you very much for improving pkgng!
